File: hackerrank/ProblemSolving/python/append-and-delete.py
#!/bin/python3

import os

def appendAndDelete(s, t, k):
  i = 0
  for x in range(min(len(s),len(t))):
    i+=1
    if s[x] != t[x]:
      break

  num=k-len(s)-len(t)+2*i

  if num<0:
    print("No")
  elif s == t:
    print("Yes")
  elif num<=2*i:
    if num%2 != 0:
      print("No")
    elif num%k == 0:
      print("No")
    else:
      print("Yes")
  else:
    print("Yes")

if __name__ == '__main__':
  fptr = open(os.environ['OUTPUT_PATH'], 'w')

  s = input()

  t = input()

  k = int(input().strip())

  result = appendAndDelete(s, t, k)

  fptr.write(result + '\n')

  fptr.close()

File: hackerrank/Algorithms/golang/append-and-delete.go
package main

import (
  "bufio"
  "fmt"
  "io"
  "os"
  "strconv"
  "strings"
  "math"
)

func appendAndDelete(s string, t string, k int32) string {
  var i int32
  for x := 0; x < int(math.Min(float64(len(s)),float64(len(t)))); x++ {
    i+=1
    if s[x] != t[x] { break; }
  }

  num := k - int32(len(s)) - int32(len(t)) + 2*i
  var result string

  if num < 0 {
    result = "No"
  } else if s == t {
    result = "Yes"
  } else if num <= 2*i {
    if num%2 != 0 {
      result = "No"
    } else if num%k == 0 {
      result = "No"
    } else { result = "Yes" }
  } else { result = "Yes" }
  
  return result
}

func main() {
  reader := bufio.NewReaderSize(os.Stdin, 16 * 1024 * 1024)

  stdout, err := os.Create(os.Getenv("OUTPUT_PATH"))
  checkError(err)

  defer stdout.Close()

  writer := bufio.NewWriterSize(stdout, 16 * 1024 * 1024)

  s := readLine(reader)

  t := readLine(reader)

  kTemp, err := strconv.ParseInt(strings.TrimSpace(readLine(reader)), 10, 64)
  checkError(err)
  k := int32(kTemp)

  result := appendAndDelete(s, t, k)

  fmt.Fprintf(writer, "%s\n", result)

  writer.Flush()
}

func readLine(reader *bufio.Reader) string {
  str, _, err := reader.ReadLine()
  if err == io.EOF { return "" }
  return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
  if err != nil { panic(err) }
}